[dino.bin] 지하철 서비스 step3 리뷰 요청합니다.

kommadot opened the pull request:
## 요약
요금 정산과 최단거리 탐색은 Util패키지의 클래스들로 계산을 합니다.

## 질문
1. Service들이 다른 서비스의 의존을 갖고있는 구조가 괜찮은 구조인가요? 그러다보면 순환참조가 발생할 수도 있을것 같아서요..
2. Util클래스는 범용적이게 구현하기보단 쓰임에 맞게 의존을 갖고있는 구조를 갖고있는데 Util패키지는 원래 그렇게 사용하나요?

감사합니다.

tttuer committed:
feat: step1, step2 반영

step1 경로 조회 기능, step2 회원 인증 관련 기능(토큰 발급 API, 즐겨 찾기 기능), 테이블 스키마 추가

tttuer committed:
feat: step3 요금 계산 기능

거리별 요금 정책 및 노선별 추가 요금 계산 후 연령 별 할인 적용.

kakao-dino-bin committed:
refactor: 유저 인증, 검증 리팩토링

Null값으로 체크 하는 방어코드를 도메인에게 명령식으로 구현

-

kakao-dino-bin committed:
refactor: PathService 리팩토링

FareUtil, ShortestUtil로 최단경로, 경로의 요금 산정 리팩토링

-

kakao-dino-bin committed:
feat: 예외처리 수정 및 추가

RuntimeException으로 되어있던 구문을 CustomException으로 변경
ControllerAdvice를 사용해서 Exception을 핸들링

-

friediceday commented on src/main/java/subway/CustomExceptionHandler.java:
```diff
@@ -0,0 +1,81 @@
+import subway.exception.*;
+
+@RestControllerAdvice
+public class CustomExceptionHandler {
```
익셉션이 종류별로 다양하게 있네요. 이렇게 상황별로 익셉션 하나씩 만드는 방법도 있지만 너무 많아지지 않을까요? 어느정도 공통되는 익셉션을 묶는 방법도 있을텐데요, 상세 메시지나 상수값으로 구분하는 방법으로요. 어떻게 생각하시나요

friediceday commented on src/main/java/subway/CustomExceptionHandler.java:
```diff
@@ -0,0 +1,81 @@
+
+    @ExceptionHandler(AlreadyExistedEmailException.class)
+    public ResponseEntity alreadyExistedEmail() {
+        return ResponseEntity.status(HttpStatus.BAD_REQUEST).build();
```
already exists는 conflict가 더 적절하지 않을까요

friediceday commented on src/main/java/subway/CustomExceptionHandler.java:
```diff
@@ -0,0 +1,81 @@
+
+    @ExceptionHandler(DuplicateNameException.class)
+    public ResponseEntity duplicateNameException() {
+        return ResponseEntity.status(HttpStatus.BAD_REQUEST).build();
```
이것도 conflict가 적절하지 않을까요

friediceday commented on src/main/java/subway/CustomExceptionHandler.java:
```diff
@@ -0,0 +1,81 @@
+
+    @ExceptionHandler(NotExistLineException.class)
+    public ResponseEntity notExistLineException() {
+        return ResponseEntity.status(HttpStatus.BAD_REQUEST).build();
```
not found가 어떨까요 아래들도요

kakao-dino-bin committed:
refactor: 자잘한 리팩토링, exception 내용 추가

kakao-dino-bin committed:
refactor: 반환값 불변으로 변경

-

friediceday commented on src/main/java/subway/favorite/domain/Favorite.java:
```diff
@@ -0,0 +1,41 @@
+import subway.exception.WrongStationIdException;
+
+public class Favorite {
+    private Long id;
```
프리미티브 대신 레퍼런스를 사용하시는 이유가 있나요

kommadot replied:
특별한 이유보단 기존의 코드가 전부 레퍼런스로 되어있어서 통일성을 주기위해서 래퍼클래스로 계속 사용했습니다.

friediceday commented on src/main/java/subway/favorite/ui/FavoriteController.java:
```diff
@@ -1,8 +1,42 @@
+        this.favoriteService = favoriteService;
+    }
+
     // TODO: 즐겨찾기 기능 구현하기
```
완료됐으면 삭제해주세요

friediceday commented on src/main/java/subway/line/dao/LineDao.java:
```diff
@@ -32,36 +34,45 @@ public Line insert(Line line) {
+        try {
+            Long lineId = insertAction.executeAndReturnKey(params).longValue();
+            return new Line(lineId, line.getName(), line.getColor());
+        } catch (RuntimeException e){
```
여기서 발생하는 모든 런타임 에러를 잡으면 duplicate가 맞나요?
보니까 이런식의 처리를 많이 하셨네요.

kakao-dino-bin committed:
test: util 패키지 유닛테스트 추가

util패키지의 정확성을 테스트해 보기 위해서 테스트 추가

-

friediceday commented on src/main/java/subway/member/domain/LoginMember.java:
```diff
@@ -25,4 +25,16 @@ public String getEmail() {
+    }
+
+    public boolean isChild() {
+        return isLoginMember() && 6 <= age && age < 13;
```
상수를 쓰면 어떨까요

friediceday commented on src/main/java/subway/member/domain/Member.java:
```diff
@@ -9,23 +15,26 @@
-        this.email = email;
+    public Member(String email, String password, Integer age) {
+        this(email, password);
+        if (age < 1) {
```
상수를 쓰면 어떨까요

ljh131 reviewed (changes requested):
안녕하세요. 테스트를 돌려보니 일부 실패가 있네요. 모든 테스트는 성공해야 합니다.
그리고 pr후 계속 수정 하셔서 리뷰가 어렵습니다. 수정이 완료되면 다시 요청해주세요.

kakao-dino-bin committed:
fix: 이메일 인증 실패 버그와 테스트코드의 실패시 기대 상태 코드 변경

멤버를 생성할때 이메일에 대한 검증을 진행하는데 정규식이 잘못되어서 수정하였습니다.
테스트 코드중 구간관련 기능 실패시 기대 StatusCode가 500이었는데 이를 400으로 수정

-

kakao-dino-bin committed:
refactor: 피드백 주신것 반영(상수화, todo삭제 등)

-

kakao-dino-bin committed:
refactor: 피드백 주신것 반영(상수화, todo삭제 등)

-

kakao-dino-bin committed:
refactor: ExceptionHandler피드백 반영

Exception의 갯수를 줄이고 인자 전달식으로 내용이 전달되도록 변경
또한 Runtime형식으로 받던 Exception을 고정된 Exception을 받도록 변경

-

kakao-dino-bin committed:
modify: bad_request반환 메소드를 not_found를 반환하도록 변경

-

kakao-dino-bin committed:
test: 테스트에도 status코드 변경사항 반영

-

friediceday commented on src/main/java/subway/path/application/PathService.java:
```diff
@@ -0,0 +1,44 @@
+        this.lineDao = lineDao;
+    }
+
+    public PathResponse findShortestPath(LoginMember loginMember, Long source, Long target) {
```
shortest path인데 여러가지 일을 하는 것 같아요. fare도 계산하고요.
이름을 좀 더 적합하게 바꾸고 메서드도 역할별로 쪼개보면 좋겠습니다

friediceday commented on src/main/java/subway/auth/application/AuthService.java:
```diff
@@ -1,4 +1,43 @@
+            memberDao.findByEmail(tokenRequest.getEmail())
+                    .checkValidMember(new Member(tokenRequest.getEmail(), tokenRequest.getPassword()));
+        } catch (IncorrectResultSizeDataAccessException e) {
+            throw new LoginFailException();
```
create token인데 로그인에 관한 익셉션을 준다면, 로그인은 메서드 역할보다 더 상위 개념이라서 어색하지 않을까요?

friediceday commented on src/main/java/subway/auth/ui/AuthenticationPrincipalArgumentResolver.java:
```diff
@@ -20,10 +25,13 @@ public boolean supportsParameter(MethodParameter parameter) {
-        return null;
+        String token = AuthorizationExtractor.extract((HttpServletRequest) webRequest.getNativeRequest());
+        if (token == null) {
+            return new LoginMember(-1L, "", -1);
```
이러면 로그인 상태가 아닌데 로그인멤버를 가지게 되나본데요. 더 좋은 방법은 없을까요

friediceday commented on src/main/java/subway/member/domain/LoginMember.java:
```diff
@@ -25,4 +28,16 @@ public String getEmail() {
         return age;
     }
+
+    public boolean isLoginMember() {
```
로그인 멤버가 아닌데 로그인 멤버일수가 있는거니, 매번 객체를 사용할 때마다 유효한지 확인이 필요하겠네요.

friediceday commented on src/main/java/subway/util/FareUtil.java:
```diff
@@ -0,0 +1,52 @@
+
+public class FareUtil {
+    private static final int BASE_FARE = 1250;
+    private static final int HUNDRED = 100;
```
이름이 100인것보다는 더 좋은 이름을 찾아보면 좋겠습니다.

friediceday commented on src/main/java/subway/member/domain/Member.java:
```diff
@@ -9,23 +16,26 @@
 
-    public Member(Long id, String email, String password, Integer age) {
-        this.id = id;
+    public Member(String email, String password) {
```
age는 없어도 되는 값인가요?

friediceday commented on src/main/java/subway/line/domain/Section.java:
```diff
@@ -12,16 +15,24 @@ public Section() {
 
     public Section(Station upStation, Station downStation, int distance) {
+        this(upStation, downStation);
+        if (distance < 1) {
```
상수를 쓰면 어떨까요

friediceday commented on src/main/java/subway/line/domain/Sections.java:
```diff
@@ -120,7 +130,7 @@ private Section findSectionByNextUpStation(Station station) {
 
     public void removeStation(Station station) {
         if (sections.size() <= 1) {
```
이건 주어진 코드였나요?

kommadot replied:
네 원래 있던 뼈대 코드부분 인것 같습니다.

ljh131 reviewed (changes requested):
잘 수정하셨습니다. 몇 가지 더 커멘트 드렸습니다.

ljh131 commented:
> ## 질문
> 1. Service들이 다른 서비스의 의존을 갖고있는 구조가 괜찮은 구조인가요? 그러다보면 순환참조가 발생할 수도 있을것 같아서요..
> 2. Util클래스는 범용적이게 구현하기보단 쓰임에 맞게 의존을 갖고있는 구조를 갖고있는데 Util패키지는 원래 그렇게 사용하나요?

1 -> 되도록 순환참조가 발생하지 않도록 하는게 좋겠습니다.

2-> 비지니스에 독립으로 범용적인게 좋겠습니다. 만약 비지니스 로직을 담고있다면 유틸이라기보단 다른 이름, 다른 패키지면 좋을 것 같아요.

kakao-dino-bin committed:
refactor: 피드백 반영(상수화, 이름변경, 멤버 역할 강조)

-

kakao-dino-bin committed:
modify: LoginModel전달이 아닌 enum클래스의 객체를 인자로해서 요금을 계산

-

kakao-dino-bin committed:
refactor: 메소드명 변경과 메소드 분리

-

kakao-dino-bin committed:
refactor: 최단거리를 구하는 Util 클래스를 Util이 아닌 도메인으로 변경

Path란 도메인 내에서 최단경로와, 요금을 구하도록 변경

-

kakao-dino-bin committed:
refactor: 경로를 구하는 클래스를 하나의 도메인으로 변경 및 테스트

-

kakao-dino-bin committed:
refactor: 패키지 이동 및 패키지 변경

-

kakao-dino-bin committed:
refactor: 패키지 이름 변경

friediceday commented on src/main/java/subway/CustomExceptionHandler.java:
```diff
@@ -0,0 +1,41 @@
+
+    @ExceptionHandler(AuthorizationFailException.class)
+    public ResponseEntity<String> authorizationFail() {
+        return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body("인증이 실패하였습니다.");
```
authentication과 authorization의 차이점을 알아보고 적절히 사용해보시면 좋겠습니다

friediceday commented on src/main/java/subway/path/domain/KakaoMap.java:
```diff
@@ -0,0 +1,46 @@
+
+import java.util.List;
+
+public class KakaoMap {
```
kakao map은 무슨 의미를 갖고있나요?

kommadot replied:
요금을 계산하고, 거리를 측정하는 클래스의 적절한 이름이 떠오르지 않아서 명명한 이름입니다. 그치만 말씀하신대로 의미를 파악하기 힘들어 다시 명명하였습니다.

friediceday commented on src/main/java/subway/path/domain/KakaoMap.java:
```diff
@@ -0,0 +1,46 @@
+    MetroMap metroMap;
+    Station source;
+    Station target;
+
```
다른데서 사용해서 패키지프라이빗으로 하신건가요?

kommadot replied:
실수로 안적어둔것 같습니다 수정하겠습니다

friediceday commented on src/main/java/subway/path/domain/ComplimentaryAge.java:
```diff
@@ -0,0 +1,41 @@
+import java.util.stream.Stream;
+
+public enum ComplimentaryAge {
+    TODDLER(0, 5, 1), CHILD(6, 12, 0.5), TEENAGER(13, 18, 0.2), ADULT(19, 64, 0), ELDER(65, 1000, 1);
```
한줄에 하나씩 쓰는게 가독성에 좋지 않을까합니다

friediceday commented on src/main/java/subway/exception/NoMoreSectionToDeleteException.java:
```diff
@@ -0,0 +1,4 @@
+package subway.exception;
+
+public class NoMoreSectionToDeleteException extends RuntimeException{
```
공백+중괄호열기 이런식으로 대부분 하셨으니 일관성을 지키도록 해보시면 좋겠습니다.

friediceday commented on src/main/java/subway/auth/ui/AuthController.java:
```diff
@@ -1,5 +1,24 @@
+    public AuthController(AuthService authService){
+        this.authService = authService;
+    }
+    @PostMapping("/token")
```
여기도 다른곳과 마찬가지로 메서드별로 한줄씩 공백을 주도록 해보시면 좋겠습니다

friediceday commented on src/main/java/subway/favorite/domain/Favorite.java:
```diff
@@ -0,0 +1,41 @@
+    }
+
+    public Favorite(Long id, Long memberId, Long sourceStationId, Long targetStationId) {
+        this(memberId, sourceStationId, targetStationId);
```
보통 작은 생성자가 큰 생성자를 호출하는게 일반적이지 않을까요. 그럼 final도 쓸수있고요

kommadot replied:
필드들을 final로 선언하고 생성자에서 초기화를 다 해준다는 말씀이군요. 그렇다면 id값처럼 아직 정해지지 않은 값에 대해선 null을 넣은채로 호출을 하면 될까요?

friediceday commented on src/main/java/subway/line/domain/Section.java:
```diff
@@ -1,8 +1,12 @@
+import java.util.Objects;
+
 public class Section {
+    public static final int MIN_DISTANCE = 1;
```
외부에서 사용하지 않는 것 같습니다.

friediceday commented on src/main/java/subway/line/domain/Sections.java:
```diff
@@ -1,18 +1,19 @@
 import java.util.stream.Collectors;
 
 public class Sections {
+    public static final int MIN_SIZE = 1;
```
이것도 여기서만 사용하는 것 같네요

friediceday commented on src/main/java/subway/path/domain/MetroMap.java:
```diff
@@ -0,0 +1,50 @@
+import java.util.stream.IntStream;
+
+public class MetroMap {
+    DijkstraShortestPath<Station, DefaultWeightedEdge> dijkstraShortestPath;
```
메트로맵은 이름이 좀 추상적인것 같아요. 하는일은 최단거리와 관련된 일만 하는데요.

friediceday commented on src/main/java/subway/path/caculator/FareCalculator.java:
```diff
@@ -0,0 +1,50 @@
+    public static int getDistanceFare(int totalDistance) {
+        int distanceFare = BASE_FARE;
+        if (BASIC_FARE_DISTANCE < totalDistance && totalDistance <= FIFTY_KILOMETER) {
+            distanceFare += Math.ceil((totalDistance - BASIC_FARE_DISTANCE) / MIDDLE_DISTANCE_DENOMINATOR) * PERCENTAGE_VALUE;
```
여기 100을 곱하는게 퍼센트라고 표현하면 적절한지 좀 의문입니다. 5km당 추가운임 아닌가요?

friediceday commented on src/main/java/subway/path/application/PathService.java:
```diff
@@ -0,0 +1,41 @@
+        this.lineDao = lineDao;
+    }
+
+    public PathResponse getPathAndFare(LoginMember loginMember, Long source, Long target) {
```
source, target은 정확히 어떤값을 의미하는지 명시하면 좋겠습니다. source station id를 의미하는거 맞나요? 로컬변수면 몰라도... 메서드 시그니쳐니까요.

friediceday commented on src/main/java/subway/line/domain/Section.java:
```diff
@@ -12,16 +16,24 @@ public Section() {
     }
 
     public Section(Station upStation, Station downStation, int distance) {
+        this(upStation, downStation);
```
여기도 더 큰 생성자를 호출하는 식이 어떨까요

friediceday commented on src/main/java/subway/member/domain/Member.java:
```diff
@@ -43,4 +49,23 @@ public String getPassword() {
+
+    public void checkValidPassword(String password) {
+        if (!this.password.equals(password)) {
+            throw new LoginFailException();
```
여기도 패스워드 체크하는데 로그인실패 에러가 나가면 별로 적절해보이지 않습니다.

friediceday commented on src/main/java/subway/line/domain/Section.java:
```diff
@@ -12,16 +16,24 @@ public Section() {
     }
 
     public Section(Long id, Station upStation, Station downStation, int distance) {
+        this(upStation, downStation, distance);
```
여기선 여러 생성자를 타고가네요

ljh131 reviewed (changes requested):
안녕하세요. 조금 더 수정해보면 좋을 것 같습니다.

kakao-dino-bin committed:
refactor: 가독성을 위한 개행 및 네이밍 변경

-

kakao-dino-bin committed:
refactor: 코딩 컨벤션에 맞게 템플릿 적용

-

kakao-dino-bin committed:
refactor: 변수명 변경, 클래스명 변경

-

kakao-dino-bin committed:
refactor: 변수명 역할에 더 일치하도록 변경

-

kakao-dino-bin committed:
refactor: 비밀번호 비교 실패시, 예외처리 위치 변경

-

kakao-dino-bin committed:
refactor:  작은 생성자가 큰 생성자를 호출하도록 변경

-

kakao-dino-bin committed:
fix: boolean결과 분기점 수정

friediceday commented on src/main/java/subway/path/application/PathService.java:
```diff
@@ -0,0 +1,41 @@
+            complimentaryAge = ComplimentaryAge.getAgeGroup(loginMember.getAge());
+        }
+        MetroNavigator metroNavigator = new MetroNavigator(lines, stationDao.findById(startStationId), stationDao.findById(destStationId), complimentaryAge);
+        return new PathResponse(metroNavigator.getShortestPath().stream()
```
이런 부분은 하나에 몰아넣지말고 분리하면 가독성에 좋을 것 같습니다.

friediceday commented on src/main/java/subway/path/caculator/FareCalculator.java:
```diff
@@ -0,0 +1,48 @@
+        if (BASIC_FARE_DISTANCE < totalDistance && totalDistance <= FIFTY_KILOMETER) {
+            distanceFare += Math.ceil((totalDistance - BASIC_FARE_DISTANCE) / MIDDLE_DISTANCE_DENOMINATOR) * EXTRA_FARE_PER_DISTANCE;
+        }
+        if (totalDistance > FIFTY_KILOMETER) {
```
else if를 써도 되지 않을까요. 나중에 if cond가 잘못되더라도 두번 누적되는 버그를 막을수도 있겠고요.

friediceday commented on src/main/java/subway/path/caculator/FareCalculator.java:
```diff
@@ -0,0 +1,48 @@
+    private static final int EXTRA_FARE_PER_DISTANCE = 100;
+    private static final int DEDUCTION_FARE = 350;
+    private static final int BASIC_FARE_DISTANCE = 10;
+    private static final int FIFTY_KILOMETER = 50;
```
상수 이름을 50km라고 하기보다는 50의 의미를 나타낼 수 있는 이름이 더 좋을 것 같습니다. 아래도요

friediceday commented on src/main/java/subway/path/caculator/FareCalculator.java:
```diff
@@ -0,0 +1,48 @@
+    private static final int OVER_FIFTY_KILOMETER_FARE = 800;
+    private static final double MIDDLE_DISTANCE_DENOMINATOR = 5;
+    private static final double HIGH_DISTANCE_DENOMINATOR = 8;
+    public static final int ZERO_FARE = 0;
```
여기도 현재 0이지만 나중에 100원을 받겠다고 했을때 이름이 그대로 zero면 적절치 않을 것 같습니다

friediceday replied:
외부에서 사용하지 않는 것 같습니다

ljh131 reviewed (commented):
수고하셨습니다. 잘 수정해주신 것 같습니다. 이번 단계는 여기서 마무리하고 머지하겠습니다.
